faucet/src/lib.rs

```diff
@@ -56,125 +63,277 @@ pub mod pallet {
 	pub struct Pallet<T>(PhantomData<T>);
 
 	#[pallet::config]
-	pub trait Config: frame_system::Config + encointer_reputation_commitments::Config {
+	pub trait Config:
+		frame_system::Config
+		+ encointer_reputation_commitments::Config
+		+ encointer_communities::Config
+		+ pallet_treasury::Config
+	{
 		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
-		type Currency: Currency<Self::AccountId>;
+		type Currency: Currency<Self::AccountId> + NamedReservableCurrency<Self::AccountId>;
 		type ControllerOrigin: EnsureOrigin<Self::RuntimeOrigin>;
-
 		#[pallet::constant]
 		type PalletId: Get<PalletId>;
 	}
 
 	#[pallet::call]
-	impl<T: Config> Pallet<T> {
-		/// Transfer some balance to another account.
+	impl<T: Config> Pallet<T>
+	where
+		sp_core::H256: From<<T as frame_system::Config>::Hash>,
+		T::AccountId: AsRef<[u8; 32]>,
+		ReserveIdentifierOf<T>: From<[u8; 8]>,
+	{
 		#[pallet::call_index(0)]
 		#[pallet::weight(10_000)]
+		pub fn create_faucet(
+			origin: OriginFor<T>,
+			name: FaucetNameType,
+			amount: BalanceOf<T>,
+			whitelist: WhiteListType,
+			drip_amount: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let from = ensure_signed(origin)?;
+
+			let all_communities = encointer_communities::Pallet::<T>::community_identifiers();
+			for cid in &whitelist {
+				if !all_communities.contains(cid) {
+					return Err(<Error<T>>::InvalidCommunityIdentifierInWhitelist.into())
+				}
+			}
+
+			// create account
+			let faucet_identifier =
+				[<T as Config>::PalletId::get().0.as_slice(), name.to_vec().as_slice()].concat();
+
+			let faucet_id_hash: H256 = T::Hashing::hash_of(&faucet_identifier).into();
+			let faucet_account = T::AccountId::decode(&mut faucet_id_hash.as_bytes())
+				.expect("32 bytes can always construct an AccountId32");
+
+			if <Faucets<T>>::contains_key(&faucet_account) {
+				return Err(<Error<T>>::FaucetAlreadyExists.into())
+			}
+
+			<T as Config>::Currency::reserve_named(
+				&Self::get_reserve_id(&faucet_account),
```

Comment:
It is dangerous if you start altering state before all verifications have been done:
If the caller has enough balance to reserve, but not enough for the initial amount specified, the funds will be reserved but the faucet won't be created and the reserved funds will be lost.

Or am I missing something?

Reply:
@clangenb confirmed to me that extrinsics are now atomic, so if anything fails in the extrinsic, the rest will be rolled back.

Reply:
a major advantage over solidity smart contracts